Source/ROSIntegrationVision/Private/VisionComponent.cpp

```diff
@@ -0,0 +1,755 @@
+        Object->TextureTarget = CreateDefaultSubobject<UTextureRenderTarget2D>(TEXT("ObjectTarget"));
+        Object->TextureTarget->InitAutoFormat(Width, Height);
+        Object->FOVAngle = FieldOfView;
+
+        UE_LOG(LogTemp, Warning, TEXT("Loading materials"))
+            ConstructorHelpers::FObjectFinder<UMaterial> MaterialDepthFinder(TEXT("Material'/ROSIntegrationVision/SceneDepth.SceneDepth'"));
+        if (MaterialDepthFinder.Object != nullptr)
+        {
+            MaterialDepthInstance = UMaterialInstanceDynamic::Create(MaterialDepthFinder.Object, Depth);
+            if (MaterialDepthInstance != nullptr)
+            {
+                Depth->PostProcessSettings.AddBlendable(MaterialDepthInstance, 1);
+            }
+        }
+        else {
+            UE_LOG(LogTemp, Error, TEXT("Could not load material for depth."));
+        }
+    }
+    else {
+        UE_LOG(LogTemp, Warning, TEXT("No owner!"));
+    }
+}
+
+UVisionComponent::~UVisionComponent()
+{
+    delete Priv;
+}
+
+void UVisionComponent::SetFramerate(const float _Framerate)
+{
+    Framerate = _Framerate;
+    FrameTime = 1.0f / _Framerate;
+    TimePassed = 0;
+}
+
+void UVisionComponent::Pause(const bool _Pause)
+{
+    Paused = _Pause;
+}
+
+bool UVisionComponent::IsPaused() const
+{
+    return Paused;
+}  
+
+void UVisionComponent::InitializeComponent()
+{
+    Super::InitializeComponent();
+}
+
+void UVisionComponent::BeginPlay()
+{
+    Super::BeginPlay();
+    // Initializing buffers for reading images from the GPU
+	ImageColor.AddUninitialized(Width * Height);
+	ImageDepth.AddUninitialized(Width * Height);
+	ImageObject.AddUninitialized(Width * Height);
+
+	// Reinit renderer
+	Color->TextureTarget->InitAutoFormat(Width, Height);
+	Depth->TextureTarget->InitAutoFormat(Width, Height);
+	Object->TextureTarget->InitAutoFormat(Width, Height);
+
+	AspectRatio = Width / (float)Height;
+
+	// Setting flags for each camera
+	ShowFlagsLit(Color->ShowFlags);
+	ShowFlagsPostProcess(Depth->ShowFlags);
+	ShowFlagsVertexColor(Object->ShowFlags);
+
+	// Creating double buffer and setting the pointer of the server object
+	Priv->Buffer = TSharedPtr<PacketBuffer>(new PacketBuffer(Width, Height, FieldOfView));
+
+	Running = true;
+	Paused = false;
+
+	Priv->DoColor = false;
+	Priv->DoObject = false;
+	Priv->DoDepth = false;
+
+	Priv->DoneColor = false;
+	Priv->DoneObject = false;
+
+	// Starting threads to process image data
+	Priv->ThreadColor = std::thread(&UVisionComponent::ProcessColor, this);
+	Priv->ThreadDepth = std::thread(&UVisionComponent::ProcessDepth, this);
+	Priv->ThreadObject = std::thread(&UVisionComponent::ProcessObject, this);
+
+	// Establish ROS communication
+	UROSIntegrationGameInstance* rosinst = Cast<UROSIntegrationGameInstance>(GetOwner()->GetGameInstance());
+	if (rosinst)
+	{
+		UE_LOG(LogTemp, Warning, TEXT("ROSIntegrationGameInstance available. Setting up ROS Topics ..."));
+		_TFPublisher = NewObject<UTopic>(UTopic::StaticClass());
```

Comment:
Here should also go a check if TF should be published. If that's not the case, the topic doesn't have to advertised.